File: src/ultraloom/gate.py
"""Approval points, read back from the journal.

A paused run has an address: the journal's last entry names the gate and the
question it asked, so `resume` needs nothing but the journal.
"""

from __future__ import annotations

from dataclasses import dataclass

from ultraloom.journal import Journal


@dataclass(frozen=True, slots=True)
class PendingGate:
    """A gate that stopped a run and is waiting for an answer."""

    node: str
    question: str


def pending_gate(journal: Journal) -> PendingGate | None:
    """The open question of this run, or None if nothing is waiting."""
    entries = journal.entries()
    if not entries:
        return None
    last = entries[-1]
    if last.outcome != "paused" or last.detail is None:
        return None
    return PendingGate(last.node, last.detail)

File: internal/journal/gate.go
package journal

// PendingGate is gate.py's PendingGate: a gate that stopped a run and is
// waiting for an answer.
//
// InputHash identifies the visit and not the node. A gate on a cycle pauses
// once per pass, and an answer addressed only to the node name would be spent
// on the first pass -- which the journal has already answered -- instead of on
// the pause that is actually open.
type PendingGate struct {
	Node      string
	Question  string
	InputHash string
}

// Pending is gate.py's `pending_gate`: the open question of this run, or nil.
//
// A damaged journal comes back as an error and never as "nothing waiting": a
// run whose journal cannot be read is not a run that has been answered.
func Pending(path string) (*PendingGate, error) {
	entries, err := Entries(path)
	if err != nil {
		return nil, err
	}
	if len(entries) == 0 {
		return nil, nil
	}
	last := entries[len(entries)-1]
	if last.Outcome != "paused" || last.Detail == nil {
		return nil, nil
	}
	return &PendingGate{Node: last.Node, Question: *last.Detail, InputHash: last.InputHash}, nil
}
